File: matlab_code/encoder_RAGFT_RLGR_large_blocks.m
%
% clear;
% addpath(genpath('RA-GFT'));
function [] = encoder_RAGFT_RLGR_large_blocks(dataset,sequence,b,colorStep,experiment)

T = get_pointCloud_nFrames(dataset, sequence );
nSteps = length(colorStep);
bytes = zeros(T,nSteps);
MSE  = bytes;
Nvox = zeros(T,1);
time = Nvox;
%


param.bsize = b;


for frame =1:T
    tic;
    [ V,Crgb,J ] = get_pointCloud(dataset, sequence,  frame );
    N = size(V,1);
    Nvox(frame) = N;
    C = RGBtoYUV(Crgb); %transform to YUV
    
    %
    %[Coeff,w]=RAHT(C,ListC,FlagsC,weightsC);
    param.V=V;
    param.J=J;
    last_idx = min(length(experiment),9);
    if(strcmp('exp_zhang',experiment(1:last_idx)))
        param.isMultiLevel=0;
    else
        param.isMultiLevel=1;
    end
    [Coeff, Gfreq, weights]  = Region_Adaptive_GFT( C, param );
    Y = Coeff(:,1);
    
    for i=1:nSteps
        %quantize coeffs
        step = colorStep(i);
        Coeff_enc = round(Coeff/step);
        Y_hat = Coeff_enc(:,1)*step;
        %U_hat = Coeff_enc(:,2)*step;
        %V_hat = Coeff_enc(:,3)*step;
        
        %comptue squared error
        MSE(frame,i) = (norm(Y-Y_hat)^2/(N*255^2));
        
        %encode coeffs using RLGR
        [nbytesY,~]=RLGR_encoder(Coeff_enc(:,1));
        [nbytesU,~]=RLGR_encoder(Coeff_enc(:,2));
        [nbytesV,~]=RLGR_encoder(Coeff_enc(:,3));
        bytes(frame,i) = nbytesY + nbytesU + nbytesV;
        
    end
    time(frame)=toc;
    
    fprintf('%s/%s/\t %s\t %f\t %d\t %d\n',dataset,sequence, experiment, time(frame),  frame, T);
    
    %disp(disp_str);
end

%
folder = sprintf('RA-GFT/results/%s/%s/',dataset,sequence);
mkdir(folder);
filename  = sprintf('%s_RA-GFT_%s.mat',folder,experiment);
save(filename,'MSE','bytes','Nvox','b','colorStep','time');

File: experiments/old/encoder_RAGFT_RLGR_large_blocks.py
import numpy as np
from scipy.linalg import norm
from scipy.io import savemat
from time import time
from raht import raht_forward
from rlgr import RLGR_encode
from utils import get_point_cloud, rgb_to_yuv

def encoder_RAGFT_RLGR_large_blocks(dataset, sequence, b, color_step, experiment):
    T = len(sequence)
    n_steps = len(color_step)
    bytes_array = np.zeros((T, n_steps))
    mse_array = np.zeros((T, n_steps))
    n_voxels = np.zeros(T)
    time_array = np.zeros(T)

    param = {'bsize': b}

    for frame in range(T):
        start_time = time()
        V, Crgb, J = get_point_cloud(dataset, sequence[frame])
        N = V.shape[0]
        n_voxels[frame] = N
        C = rgb_to_yuv(Crgb)

        if experiment.startswith('exp_zhang'):
            param['isMultiLevel'] = False
        else:
            param['isMultiLevel'] = True

        Coeff, Gfreq, weights = raht_forward(C, param)
        Y = Coeff[:, 0]

        for i, step in enumerate(color_step):
            # Quantize coefficients
            Coeff_enc = np.round(Coeff / step)
            Y_hat = Coeff_enc[:, 0] * step

            # Compute squared error
            mse_array[frame, i] = norm(Y - Y_hat)**2 / (N * 255**2)

            # Encode coefficients using RLGR
            nbytes_Y, _ = RLGR_encode(Coeff_enc[:, 0])
            nbytes_U, _ = RLGR_encode(Coeff_enc[:, 1])
            nbytes_V, _ = RLGR_encode(Coeff_enc[:, 2])
            bytes_array[frame, i] = nbytes_Y + nbytes_U + nbytes_V

        end_time = time()
        time_array[frame] = end_time - start_time

        print(f"{dataset}/{sequence[frame]}/\t {experiment}\t {time_array[frame]}\t {frame}\t {T}")

    folder = f"RA-GFT/results/{dataset}/{sequence}/"
    filename = f"{folder}RA-GFT_{experiment}.mat"
    savemat(filename, {'MSE': mse_array, 'bytes': bytes_array, 'Nvox': n_voxels,
                       'b': b, 'colorStep': color_step, 'time': time_array})

# Example usage:
dataset = 'example_dataset'
sequence = ['sequence1', 'sequence2', 'sequence3']  # Example sequence names
b = 8  # Block size
color_step = [0.1, 0.2, 0.5]  # Example color step values
experiment = 'exp_zhang_example'  # Example experiment name

encoder_RAGFT_RLGR_large_blocks(dataset, sequence, b, color_step, experiment)
